design/parallel-commit/parallel-commit-solution-ideas.md

```diff
@@ -0,0 +1,59 @@
+
+For point-get operations, access latch to see if the key is being prewritten with a smaller ts. If so, block or return KeyIsLocked err. For range scanning operations, check each items in `V` to see if their range overlaps. If so, block or return KeyIsLocked err.
+
+### Some other ideas
+
+* Make something like HLC, which doesn't seem to be practical for us.
+* ...
+
+## Replica Read (By @gengliqi)
+
+In the solutions to the locking problem, writing are performed on leaders, and it needs to know the `max_read_ts`. However if follower read read is enabled, the leader need to know the `max_read_ts` among all replicas of the Region by some way, and reading on followers should be blocked when the leader has an ongoing conflicting prewrite. Here is one possible solution to this problem.
+
+The main idea is to adjust the lock's `max_read_ts` in Raft layer. It might be hard to avoid Raftstore coupling with transaction logic though.
+
+First, readings on followers should send the read ts via the ReadIndex message to the leader, and the leader records it. When a prewrite of a Parallel-Commit transaction is being proposed in Raft layer, it's `max_read_ts` field should be updated if it's smaller than the `max_read_ts` that was recorded in Raft layer. This makes it possible to write down the `max_read_ts` among all replicas.
+
+However this doesn't apply to the case that a ReadIndex arrives between a prewrite's proposal and committing. We can't edit `max_read_ts` from the raft log since it has been proposed, and we cannot immediately allow the follower to read before the log being applied. So secondly, we need a additional mechanism to prevent the follower from reading without the lock. 
+
+In the current implementation (where Parallel Commit is not supported), ReadIndex returns the leader's commit index, and the follower can read only when its apply index >= leader's commit index. So one way to solve this problem, is to let the leader return `max{commit_index, pc_index}` where `pc_index` indicates the index of proposed prewrite of Parallel-Commit transactions. This is easy to implement, but increases latency of follower read, since the follower needs to wait for applying more logs before it's permitted to read. Another approach is to let the leader returns both `commit_index` and `pc_index` for ReadIndex, and the follower needs to construct a in-memory lock for committed-but-not-yet-applied prewrites, and reads are permitted when the leader's `commit_index` is applied and, additionally, `pc_index` has been committed. Then the read should be blocked if it tries to read keys that has been locked by the in-memory lock we just mentioned.
+
+Note that Raft learners also need to do this. TiFlash might be exclusive with parallel commit before we support this mechanism on it.

```

Comment:
```suggestion
In the current implementation (where Parallel Commit is not supported), ReadIndex returns the leader's commit index, and the follower can read only when its apply index >= leader's commit index. 
One way to solve this problem, is to let the leader returns `pc_index` after this `pc_index` log is committed if `pc_index` is greater than the commit index. `pc_index` indicates the index of proposed prewrite of Parallel-Commit transactions. This is easy to implement, but increases the latency of follower read, since the follower needs to wait for applying more logs before it's permitted to read. 
Another approach is to let the leader returns both `commit_index` and `pc_index` for ReadIndex, and the follower needs to construct a in-memory lock for received-but-not-yet-committed prewrites, and reads are permitted when the leader's `commit_index` is applied and, additionally, all logs before `pc_index` has been received. Then the read should be blocked if it tries to read keys that has been locked by the in-memory lock we just mentioned. Note that Raft learners also need to do this. If we choose this approach, TiFlash might be exclusive with parallel commit before we support this mechanism on it.
```